apps/client/src/features/app-settings/panel/sources-panel/useGoogleSheet.ts

```diff
@@ -0,0 +1,145 @@
+import { ChangeEvent } from 'react';
+import { useQueryClient } from '@tanstack/react-query';
+import { OntimeRundown, UserFields } from 'ontime-types';
+import { ExcelImportMap } from 'ontime-utils';
+
+import { RUNDOWN, USERFIELDS } from '../../../../common/api/apiConstants';
+import { maybeAxiosError } from '../../../../common/api/apiUtils';
+import {
+  getAuthentication,
+  getSheetsAuthUrl,
+  patchData,
+  postId,
+  postPreviewSheet,
+  postPushSheet,
+  postWorksheet,
+  uploadSheetClientFile,
+} from '../../../../common/api/ontimeApi';
+import { openLink } from '../../../../common/utils/linkUtils';
+
+import { useSheetStore } from './useSheetStore';
+
+// TODO: recover useEffect for resuming previous state
+export default function useGoogleSheet() {
+  const queryClient = useQueryClient();
+
+  // functions push data to store
+  const setClientSecret = useSheetStore((state) => state.setClientSecret);
+  const patchStepData = useSheetStore((state) => state.patchStepData);
+  const setWorksheetOptions = useSheetStore((state) => state.setWorksheetOptions);
+  const setRundown = useSheetStore((state) => state.setRundown);
+  const setUserFields = useSheetStore((state) => state.setUserFields);
+
+  /** receives a client secrets file and passes on to the server */
+  const handleClientSecret = async (event: ChangeEvent<HTMLInputElement>) => {
+    if (!event.target.files?.length) {
+      patchStepData({
+        clientSecret: { available: true, error: 'Missing file' },
+        authenticate: { available: false, error: '' },
+      });
+      return;
+    }
+
+    try {
+      const selectedFile = event.target.files[0];
+      await uploadSheetClientFile(selectedFile);
+      setClientSecret(selectedFile);
+      patchStepData({
+        clientSecret: { available: true, error: '' },
+        authenticate: { available: true, error: '' },
+      });
+    } catch (error) {
+      patchStepData({
+        clientSecret: { available: true, error: maybeAxiosError(error) },
+        authenticate: { available: false, error: '' },
+      });
+    }
+  };
+
+  /** authenticate with the Google Sheets API */
+  const handleAuthenticate = async () => {
+    try {
+      const authLink = await getSheetsAuthUrl();
+
+      // request windown to open link and check auth when user is back
+      openLink(authLink);
+      window.addEventListener('focus', async () => await getAuthentication(), { once: true });
```

Comment:
When getting the authentication status here you need to handle the result to then enable or disable the next step